yare-core/src/main/java/com/sabre/oss/yare/core/internal/DefaultEngineController.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * MIT License
+ *
+ * Copyright 2018 Sabre GLBL Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package com.sabre.oss.yare.core.internal;
+
+import com.sabre.oss.yare.core.EngineControllerEvent;
+import com.sabre.oss.yare.core.EngineControllerEventType;
+import com.sabre.oss.yare.core.EngineControllerObservable;
+import com.sabre.oss.yare.core.EngineListener;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+
+public class DefaultEngineController implements EngineControllerObservable {
+    private final List<EngineListener> listeners = new CopyOnWriteArrayList<>();
+
+    @Override
+    public void stopProcessing() {
+        for (EngineListener listener : listeners) {
+            listener.onEvent(EngineControllerEvent.of(EngineControllerEventType.STOP_PROCESSING_ALL_FACTS));
+        }
+    }
+
+    @Override
+    public void register(EngineListener listener) {
```

Comment:
I am thinking about `Listener` interface which would be just a markup for all listeners (it would not have any method). `StopProcessingListener` would extend this interface. Then this method would do a little magic (like in Spring ;) ) and would check what type of `Listener` it is and put it to appropriate listener collection. I think that similar approach you can find in JUnit 5 with `Extension` interface - all kind of callbacks for extension are extending this interface and the engine knows with which of them it is dealing and engine is handling them appropriately.

In other words in this class would be one `register(Listener)` method and many collections (e.g. `List<StopProcessingListener>` which would be filled appropriately by this method. Then user could implement all listeners in one class and pass it to `register` method - and all of listeners would be added to all collections. Of course all `EngineController` methods would call appropriate methods on those listeners.

Reply:
@wojciechzachwieja Please take a look on a short example below. It is not exact implementation - just some approximation showing the idea.

```java
public interface Listener {
}

interface StopProcessingListener extends Listener {
    void onStopProcessing(StopProcessingEvent event);
}

interface StopProcessingEvent {
    boolean shouldStopOnlyGroupedFacts();
}

interface NewFeatureListener {
    String onNewFeature(NewFeatureEvent event);
}

interface NewFeatureEvent {
    String getFeatureArgument();
}

interface EngineController {
    void stopProcessing();
    void stopProcessingGroupedFacts();

    String calculateString(String argument);
}

class DefaultEngineController implements EngineController {
    private final AtomicReference<StopProcessingListener> stopProcessingListener = new AtomicReference<>();
    private final AtomicReference<NewFeatureListener> newFeatureListener = new AtomicReference<>();

    @Override
    public void stopProcessing() {
        StopProcessingListener listener = stopProcessingListener.get();
        if (listener != null) {
            listener.onStopProcessing(() -> false);
        }
    }

    @Override
    public void stopProcessingGroupedFacts() {
        StopProcessingListener listener = stopProcessingListener.get();
        if (listener != null) {
            listener.onStopProcessing(() -> true);
        }
    }

    @Override
    public String calculateString(String argument) {
        NewFeatureListener listener = newFeatureListener.get();
        if (listener != null) {
            return listener.onNewFeature(() -> argument);
        }
        throw new NoListenerException("Missing Listener for return value creation");
    }

    public void register(Listener listener) {
        tryToRegister(StopProcessingListener.class, listener, stopProcessingListener);
        tryToRegister(NewFeatureListener.class, listener, newFeatureListener);
    }

    private <T> void tryToRegister(Class<T> clazz, Listener listener, AtomicReference<T> destination) {
        if (listener.getClass().isAssignableFrom(clazz)) {
            boolean wasUninitialized = destination.compareAndSet(null, (T) listener);
            if (!wasUninitialized) {
                throw new AlreadyRegisteredListenerException();
            }
        }
    }

    public static class AlreadyRegisteredListenerException extends RuntimeException {
    }

    public static class NoListenerException extends RuntimeException {
        public NoListenerException(String message) {
            super(message);
        }
    }
}
```

Reply:
Since it looks a little complicated I have also made a version for current requirements:

```java
public interface Listener {
}

interface StopProcessingListener extends Listener {
    void onStopProcessing(StopProcessingContext context);
}

interface StopProcessingContext {
}


interface EngineController {
    void stopProcessing();
}

class DefaultEngineController implements EngineController {
    private final AtomicReference<StopProcessingListener> stopProcessingListener = new AtomicReference<>();

    @Override
    public void stopProcessing() {
        execute(stopProcessingListener, StopProcessingListener::onStopProcessing, new StopProcessingContext() {});
    }

    public void register(Listener listener) {
        tryToRegister(StopProcessingListener.class, listener, stopProcessingListener);
    }

    private <T, C> void execute(AtomicReference<T> listener, BiConsumer<T, C> callback, C context) {
        T l = listener.get();
        if (l != null) {
            callback.accept(l, context);
        }
    }

    private <T> void tryToRegister(Class<T> clazz, Listener listener, AtomicReference<T> destination) {
        if (listener.getClass().isAssignableFrom(clazz)) {
            boolean wasUninitialized = destination.compareAndSet(null, (T) listener);
            if (!wasUninitialized) {
                throw new AlreadyRegisteredListenerException();
            }
        }
    }

    public static class AlreadyRegisteredListenerException extends RuntimeException {
    }
}
```

Reply:
Great idea overall. I would only consider using some sort of thread safe Multimap for storing listeners. It would be easier to add new one. Apart from that, I think we should go on with this concept.

Reply:
@adamszadkowski thank you for good example.  @sobolewskikamil  I used ConcurrentHashMap. I agree with you in the future it will be easy to extend this controller. I mean adding new listeners.